File: Medium/823. Binary Trees With Factors/823. Binary Trees With Factors.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int numFactoredBinaryTrees(int[] arr) {
        final int MOD = 1_000_000_007;
        int n = arr.length;
        long[] dp = new long[n];
        
        Arrays.sort(arr);
        Map<Integer, Integer> index = new HashMap<>();
        for (int i = 0; i < n; i++) {
            index.put(arr[i], i);
        }
        
        Arrays.fill(dp, 1);
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[i] % arr[j] == 0) {
                    int factor = arr[i] / arr[j];
                    if (index.containsKey(factor)) {
                        int k = index.get(factor);
                        dp[i] = (dp[i] + dp[j] * dp[k]) % MOD;
                    }
                }
            }
        }
        
        long result = 0;
        for (long count : dp) {
            result = (result + count) % MOD;
        }
        
        return (int) result;
    }
}

File: Medium/823. Binary Trees With Factors/823. Binary Trees With Factors.py
class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        MOD = 10**9 + 7
        n = len(arr)
        arr.sort()

        dp = [1] * n
        index = {}

        for i in range(n):
            index[arr[i]] = i

        for i in range(n):
            for j in range(i):
                if arr[i] % arr[j] == 0:
                    factor = arr[i] // arr[j]
                    if factor in index:
                        k = index[factor]
                        dp[i] = (dp[i] + dp[j] * dp[k]) % MOD

        result = 0
        for count in dp:
            result = (result + count) % MOD

        return result
